src/dfx/src/lib/migrate.rs

```diff
@@ -0,0 +1,116 @@
+use anyhow::{anyhow, Context, Error};
+use candid::{CandidType, Deserialize, Principal};
+use ic_agent::Identity as _;
+use ic_utils::{
+    call::AsyncCall,
+    interfaces::{
+        management_canister::{
+            builders::{CanisterSettings, InstallMode},
+            StatusCallResult,
+        },
+        ManagementCanister, WalletCanister,
+    },
+    Argument,
+};
+
+use crate::{
+    lib::waiter::waiter_with_timeout,
+    util::{assets::wallet_wasm, expiry_duration},
+};
+
+use super::{
+    environment::Environment,
+    error::DfxResult,
+    identity::{Identity, IdentityManager},
+    models::canister_id_store::CanisterIdStore,
+    network::network_descriptor::NetworkDescriptor,
+    root_key::fetch_root_key_if_needed,
+};
+
+pub async fn migrate(env: &dyn Environment, network: &NetworkDescriptor, fix: bool) -> DfxResult {
+    fetch_root_key_if_needed(env).await?;
+    let agent = env
+        .get_agent()
+        .expect("Could not get agent from environment");
+    let mut mgr = IdentityManager::new(env)?;
+    let ident = mgr.instantiate_selected_identity()?;
+    let wallet = Identity::wallet_canister_id(env, network, ident.name())
+        .map_err(|_| anyhow!("No wallet found; nothing to do"))?;
+    let wallet = WalletCanister::create(agent, wallet)
+        .await
+        .context("No accessible wallet found; nothing to do")?;
+    let mgmt = ManagementCanister::create(agent);
+    if !wallet.version_supports_u128_cycles() {
+        if fix {
+            println!("Upgrading wallet... ");
+            let wasm = wallet_wasm(env.get_logger())?;
+            mgmt.install_code(wallet.canister_id_(), &wasm)
+                .with_mode(InstallMode::Upgrade)
+                .call_and_wait(waiter_with_timeout(expiry_duration()))
+                .await
+                .context("Could not upgrade wallet")?;
+        } else {
+            println!("The wallet is outdated; run `dfx wallet upgrade`");
+        }
+    }
+    let store = CanisterIdStore::for_env(env)?;
+    for name in store.ids.keys() {
+        if let Some(id) = store.find(name) {
+            if mgmt
+                .canister_status(&id)
+                .call_and_wait(waiter_with_timeout(expiry_duration()))
+                .await
+                .is_err()
+            {
+                if fix {
+                    println!(
```

Comment:
It looks like we're assuming that any failure of the `canister_status` call means that the current identity isn't a controller of the canister in question.

What would you think about instead using the equivalent of `dfx canister info` (callable by anyone) and explicitly checking the controller list in order to determine if we should add the current identity as a controller?

Reply:
Makes sense, I had forgotten you could query it without the mgmt canister's method.